### ISSUE TITLE
Why you insert the real world data into ms.sac_state.replay_buffe?

### ISSUE THREAD

**Issue by hdadong** [CLeARoboticsLab/ssrl#10]
The code is here: https://github.com/CLeARoboticsLab/ssrl/blob/main/ssrl_hardware/ssrl_ros_go1/scripts/train.py#L485
I wonder why you insert the real world data into the ms.sac_state.replay_buffer?
In simulation experiments, the ms.sac_state.replay_buffer only store the hallucination rollout generated by the model, not the environment rollout.

**Comment by hdadong** [comment]
After careful check, I find that in simulation experiments, the ms.sac_state.replay_buffer will also store the newest environment rollout  (though clear_model_buffer_after_model_train is True) and use it to update the sac policy, which is not align with the parameter proposed (real to synthetic data ratio=0.06) in the paper. 

Hope to response. Thanks!
The code is here: https://github.com/CLeARoboticsLab/ssrl/blob/main/ssrl/brax/training/agents/ssrl/train.py#L926

**Comment by hdadong** [comment]
Another question, why the clear_model_buffer_after_model_train is set to False in the hardware experiments. This mean that it will use all the data collect in the real world to update the sac policy and not the real to synthetic data ratio=0.06. 

How much data would be load by execute the load_rollout function? Does it load all the data collected in the hardware experiments?

In the hardware experiments, how many times would be need to execute the /ssrl/ssrl_hardware/ssrl_ros_go1/scripts/train.py to train the policy?

**Comment by jake-levy** [comment]
You're right, the model buffer does contain real-world data, alongside the hallucinated data, however only 1000 real-world transitions are collected per epoch, while 400,000 to 8,000,000 hallucinated transitions are added each epoch. Therefore, the probability of sampling a real-world transition from the model buffer is only 0.25% ~ 0.01%. This probability is very low and has a negligible impact on the 0.06 ratio. The reason I add real-world transitions to the model (SAC) buffer is so that I can run SAC with the same code (by setting start_model_horizon and end_model_horizon = 0 and model_training_max_epochs=0).

Each epoch is structured as follows: train model -> (hallucinate -> SAC update) -> (hallucinate -> SAC update) -> ... -> (hallucinate -> SAC update). Setting clear_model_buffer_after_model_train=true simply changes it to: train model -> clear model buffer -> (hallucinate -> SAC update) -> (hallucinate -> SAC update) -> ... -> (hallucinate -> SAC update). All this does (if set to true) is prevent hallucinated data from the previous epoch being used in the current epoch.  This setting didn't make much of a difference, and I don't remember why it is set differently in sim vs reality; I probably forgot to change it back in sim. But I can confirm that the 0.06 ratio is being used for both simulation and reality. I.e this is not true: "it will use all the data collect in the real world to update the sac policy and not the real to synthetic data ratio=0.06."

Load rollout loads only the latest real-world rollout (1000 transitions in my case) and then inserts these transitions in the env (and model) buffers. 

The /ssrl/ssrl_hardware/ssrl_ros_go1/scripts/train.py script is executed once per epoch. So the following is repeated for the desired number of epochs (18 in my case): collect data -> execute /ssrl/ssrl_hardware/ssrl_ros_go1/scripts/train.py. The instructions for this are in the "Running Experiments" section of https://github.com/CLeARoboticsLab/ssrl/blob/main/ssrl_hardware/README.md. 

I hope this answers all of your questions!
